node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -205,16 +235,102 @@ impl OrderingProvider {
 			self.included_candidates.remove(&finalized_candidate);
 		}
 	}
+
+	/// Returns ancestors of `head` in the descending order, stopping
+	/// either at the block present in cache or the latest finalized block.
+	///
+	/// Suited specifically for querying non-finalized chains, thus
+	/// doesn't rely on block numbers.
+	///
+	/// Both `head` and last are **not** included in the result.
+	async fn get_block_ancestors<Sender: SubsystemSender>(
+		&mut self,
+		sender: &mut Sender,
+		mut head: Hash,
+		mut head_number: BlockNumber,
+	) -> Result<Vec<Hash>> {
+		const ANCESTRY_STEP: usize = 10;
+		const ANCESTRY_SIZE_LIMIT: usize = 1000;
+
+		let mut ancestors = Vec::new();
+
+		if self.last_observed_blocks.get(&head).is_some() {
+			return Ok(ancestors)
+		}
+
+		let finalized_block_number = get_finalized_block_number(sender).await?;
+
+		loop {
+			let (tx, rx) = oneshot::channel();
+			let hashes = {
+				sender
+					.send_message(
+						ChainApiMessage::Ancestors {
+							hash: head,
+							k: ANCESTRY_STEP,
+							response_channel: tx,
+						}
+						.into(),
+					)
+					.await;
+
+				rx.await.map_err(NonFatal::Oneshot)?.map_err(NonFatal::ChainApi)?
```

Comment:
Better use `Fatal::ChainApiSenderDropped` here - that generic `Oneshot` is totally useless and should get cleaned up at some point.